### PR TITLE
[Feature]support short circuit read

### PR DESCRIPTION
**Pull request by Jay-ju**
Reference : https://github.com/StarRocks/starrocks/pull/32508
Why I'm doing:

What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSetType.java [hunk 1]
```diff
@@ -35,5 +35,6 @@ public enum RuleSetType {
     SINGLE_TABLE_MV_REWRITE,
     MULTI_TABLE_MV_REWRITE,
     PRUNE_EMPTY_OPERATOR,
-    NUM_RULE_SET
+    NUM_RULE_SET,
+    SHORT_CIRCUIT_SET
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSetType.java:39]:
```suggestion
    SHORT_CIRCUIT_SET,
    NUM_RULE_SET
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java [hunk 2]
```diff
@@ -39,6 +39,9 @@ public CastToEmptyRule() {
 
     @Override
     public boolean check(OptExpression input, OptimizerContext context) {
+        if (!(input.getOp() instanceof LogicalOperator)) {
+            return false;
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/CastToEmptyRule.java:44]:
```suggestion

```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java [hunk 3]
```diff
@@ -126,6 +127,9 @@ private boolean match(Pattern pattern, OptExpression root) {
     }
 
     private void deriveLogicalProperty(OptExpression root) {
+        if (root.getOp() instanceof PhysicalOperator) {
+            return;
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java:132]:
```suggestion
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 4]
```diff
@@ -303,10 +308,21 @@ private void pruneTables(OptExpression tree, TaskContext rootTaskContext, Column
     private OptExpression logicalRuleRewrite(ConnectContext connectContext,
                                              OptExpression tree,
                                              TaskContext rootTaskContext) {
+        Boolean isShortCircuit = tree.getShortCircuit();
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:311]: you can do all modify before rewriteAndValidatePlan in optimizeByCost, and refactor to a method

Comment (Jay-ju) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:311]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java [hunk 5]
```diff
@@ -96,10 +96,6 @@ private void rewrite(OptExpression parent, int childIndex, OptExpression root) {
     }
 
     private boolean match(Pattern pattern, OptExpression root) {
-        if (!pattern.matchWithoutChild(root)) {
-            return false;
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java:100]: could you explain the reason of deleting this?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 6]
```diff
@@ -303,9 +309,14 @@ private void pruneTables(OptExpression tree, TaskContext rootTaskContext, Column
...
                                              TaskContext rootTaskContext) {
-        tree = OptExpression.create(new LogicalTreeAnchorOperator(), tree);
+        Optional<OptExpression> result = ruleRewriteForShortCircuit(tree, rootTaskContext);
+        if (result.isPresent()) {
+            return result.get();
+        }
+
         ColumnRefSet requiredColumns = rootTaskContext.getRequiredColumns().clone();
         deriveLogicalProperty(tree);
+        deriveLogicalProperty(tree);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:319]: redudent code?

File: fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            return new ShortCircuitPlannerHybrid.LogicalPlanChecker(allowFilter, allowLimit, allowProject,
+                    allowSort, predicate, orderByColumns, limit);
+        } else {
+            return new BaseLogicalPlanChecker();
+        }
+    }
+
+    public static OptExpression checkSupportShortCircuitRead(OptExpression root, ConnectContext connectContext) {
+        if (!connectContext.getSessionVariable().isEnableShortCircuit()) {
+            root.setShortCircuit(false);
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java:55]: missing return here?

File: fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    }
+
+    /**
+     * Checker for select plan:
+     * 1. Pattern: Limit - Sort (TopN) - Project - Filter - Scan.
+     * 2. Scan: single row store table.
+     * 3. Filter: contains all keys in table primary keys.
+     * 4. Limit: no offset.
+     */
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java:93]: Should we process limitOperator here?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java (hunk 5 above)
Comment (Jay-ju) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteTreeTask.java:100]: sorry, code merge error

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 6 above)
Comment (Jay-ju) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:319]: delete

File: fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java (hunk 7 above)
Comment (Jay-ju) [reply fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java:55]: yes

File: fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java (hunk 8 above)
Comment (Jay-ju) [reply fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java:93]: currently can't process limit operator for column with row， comment is redundant

**Review by packy92 — DISMISSED** [review]

**Review by Seaven — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by decster** [comment]
@mergify backport branch-3.2

**Pull request merged**